### ISSUE TITLE
TPU Execution for Occupancy Flow Tutorial

### ISSUE THREAD

**Issue by mosicr** [waymo-research/waymo-open-dataset#444]
Hi, 
I am trying to run [Occupancy Flow Tutorial](https://github.com/waymo-research/waymo-open-dataset/blob/master/tutorial/tutorial_occupancy_flow.ipynb) on TPU. 
with strategy.scope():
356  while step < num_steps_to_train:
357   # Iterate over batches of the dataset.
358   inputs = next(it)
359   loss_value = train_step(inputs)
360
361 # Log every 10 batches.
362   if step % 10 == 0:
363     float_loss = float(loss_value)
364     print(f'Training loss after step {step}: {float_loss:.4f}')
365
366   step += 1

It fails with :
4859ValueError: Trying to create optimizer slot variable under the scope for tf.distribute.Strategy (<tensorf\
    low.python.distribute.tpu_strategy.TPUStrategyV2 object at 0x7fee4d7614f0>), which is different from the \
    scope used for the original variable (<tf.Variable 'conv1_conv/kernel:0' shape=(7, 7, 23, 64) dtype=float\
    32, numpy=

**Comment by rezama** [comment]
Does it make a different if you use the same scope around earlier operations as well (when the timestep_grids and waypoints) are being created?

In case you have already solved it, can you share what fixed it for you?

**Comment by mosicr** [comment]
No luck so far. I experimented with scopes and @tf.function across the original code. 
I don't think it is possible to simply take the original code and add scope / decorate with @tf.function to generate a proper parallelizable / TPU executable code.  Autograph is still not at the stage where you can do it.
For example, if I decorate the original train_step function ( training should run on TPU )  with @tf.function it fails with:
931  File "/usr/local/lib/python3.8/dist-packages/tensorflow/python/eager/function.py", line 56\
   6, in call
932    raise ValueError(
933ValueError: Arguments and signature arguments do not match. got: 764, expected: 768

That means that autograph generates code and then later on catches the error in code it generated. This is probably a question for Autograph Team ( Dan Moldovan ).
